### PR TITLE
(PUP-10815) Add `puppet facts diff` command 

### PR DESCRIPTION
**Pull request by Dorin-Pleava**

### CODE REVIEW

File: lib/puppet/face/facts.rb [hunk 1]
```diff
@@ -1,5 +1,24 @@
...
+partitions\..*\.uuid physicalprocessorcount processorcount processors\.count
+processors\.physicalcount selinux swapfree_mb swapsize_mb system_uptime\.days system_uptime\.hours
+system_uptime\.seconds uptime_days uptime_hours uptime_seconds clientnoop swapfree
+disks\..*\.size_bytes hypervisors\.lpar\.partition_number mountpoints\..*\.capacity
+processors\.speed serialnumber hypervisors\.xen\.privileged os\.release.\minor
+operatingsystemrelease os\.release\.full os\.distro\.description filesystems
+sp_uptime system_profiler\.uptime os\.release\.minor
+hypervisors\.zone\..* system_uptime\.uptime uptime hypervisors\.ldom\..* ldom_.*
+boardassettag dmi\.board\.asset_tag is_virtual kernelmajversion lsbmajdistrelease zones virtual
+blockdevice_.*_vendor blockdevice_.*_size]
```
Comment (gimmyxd) [inline lib/puppet/face/facts.rb:21]: I don't think we should be skipping facts that are not volatile, eq: kernelmajversion. Both cfacter and facter 4 should return the same value for that facts.

**Comment by Dorin-Pleava** [comment]
Output when no facts differ:
```
[root@dental-township ~]# puppet facts diff
2021-01-13 13:04:01.984689 WARN  puppetlabs.facter - locale environment variables were bad; continuing with LANG=C LC_ALL=C
{}
```
Output when facts differ:
[root@dental-township ~]# puppet facts diff
```
2021-01-13 13:10:05.164558 WARN  puppetlabs.facter - locale environment variables were bad; continuing with LANG=C LC_ALL=C
{"facterversion":{"new_value":"\"4.0.43\"","old_value":"\"3.14.14\""},"memory.system.available":{"new_value":"\"5.94 GiB\"","old_value":"\"6.12 GiB\""},"memory.system.available_bytes":{"new_value":"6377148416","old_value":"6572556288"},"memory.system.capacity":{"new_value":"\"22.25%\"","old_value":"\"19.87%\""}}
```
Prettier puppet facts diff with jq:
```
[root@dental-township ~]# puppet facts diff | jq
2021-01-13 13:09:44.938110 WARN  puppetlabs.facter - locale environment variables were bad; continuing with LANG=C LC_ALL=C
{
  "facterversion": {
    "new_value": "\"4.0.43\"",
    "old_value": "\"3.14.14\""
  },
  "memory.system.available": {
    "new_value": "\"5.94 GiB\"",
    "old_value": "\"6.12 GiB\""
  },
  "memory.system.available_bytes": {
    "new_value": "6376570880",
    "old_value": "6571905024"
  },
  "memory.system.capacity": {
    "new_value": "\"22.26%\"",
    "old_value": "\"19.88%\""
  }
}```

File: lib/puppet/face/facts.rb [hunk 2]
```diff
@@ -1,5 +1,23 @@
 require 'puppet/indirector/face'
 require 'puppet/node/facts'
+require 'puppet/util/fact_dif'
```
Comment (luchihoratiu) [inline lib/puppet/face/facts.rb:3]: nit: Since the action is also `diff` with double f, I would name the util file the same way.

**Review by luchihoratiu — APPROVED** [review]
Besides small comment above, it mostly LGTM! 👍 
Tested on a Ubuntu 20.04 and seems to be working as expected:
`> puppet facts diff`
> {"lsbmajdistrelease":{"new_value":"\"20\"","old_value":"\"20.04\""}}

`> puppet facts diff | jq`
> {
>   "lsbmajdistrelease": {
>     "new_value": "\"20\"",
>     "old_value": "\"20.04\""
>   }
> }

`> puppet facts diff --render-as json`
>{"lsbmajdistrelease":{"new_value":"\"20\"","old_value":"\"20.04\""}}

`> puppet facts diff --render-as yaml`
> \-\-\-
> lsbmajdistrelease:
>   :new_value: '"20"'
>   :old_value: '"20.04"'

I say mostly because I'm not sure if it's intentional for numerical values to be quoted strings (`"\"20\""` instead of `"20"`) and if `--render-as json` shouldn't prettify the output the same way `| jq` does.

File: lib/puppet/face/facts.rb [hunk 3]
```diff
@@ -1,5 +1,23 @@
...
+  memoryfree memoryfree_mb memorysize_mb mountpoints\..* mtu_.*
+  networking\.interfaces\..*\.mtu networking\.mtu partitions\..*\.filesystem
+  partitions\..*\.size_bytes partitions\..*\.mount partitions\..*\.uuid
+  system_uptime\.days system_uptime\.hours system_uptime\.seconds uptime_days
+  uptime_hours uptime_seconds disks\..*\.size_bytes
+  hypervisors\.lpar\.partition_number mountpoints\..*\.capacity
+  processors\.speed hypervisors\.xen\.privileged sp_uptime
+  system_profiler\.uptime hypervisors\.zone\..* system_uptime\.uptime uptime
+  hypervisors\.ldom\..* ldom_.* boardassettag dmi\.board\.asset_tag
+  blockdevice_.*_vendor blockdevice_.*_size is_virtual]
```
Comment (joshcooper) [inline lib/puppet/face/facts.rb:20]: Might be helpful to wrap based on different fact groups, like:

```ruby
%w[
  facterversion
  swapfree_mb swapsize_mb
  load_averages\.15m ...
  ...
```

File: lib/puppet/face/facts.rb [hunk 4]
```diff
@@ -1,5 +1,23 @@
 require 'puppet/indirector/face'
 require 'puppet/node/facts'
+require 'puppet/util/fact_dif'
+
+EXCLUDE_LIST = %w[facterversion swapfree_mb swapsize_mb load_averages\.15m
+  load_averages\.1m load_averages\.5m memory\.swap\.available_bytes
```
Comment (joshcooper) [inline lib/puppet/face/facts.rb:6]: can this be simplified to `load_averages.*`?

File: lib/puppet/face/facts.rb (hunk 3 above)
Comment (joshcooper) [inline lib/puppet/face/facts.rb:20]: Can we include `is_virtual`?

File: lib/puppet/util/fact_dif.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  def compare(fact_path, old_value)
+    new_value = @next_facter.dig(*fact_path)
+    if different?(new_value, old_value) && !excluded?(fact_path.join('.'))
+      @diff[fact_path.join('.')] = { new_value: new_value, old_value: old_value }
+    end
+  end
+
+  def different?(new, old)
+    if old.is_a?(String) && new.is_a?(String)
+      diff = old.split(',') - new.split(',')
```
Comment (joshcooper) [inline lib/puppet/util/fact_dif.rb:44]: should we also check if `new` contains entries that `old` doesn't?

File: lib/puppet/util/fact_dif.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      return true if diff.any?
+
+      return false
+    end
+
+    old != new
+  end
+
+  def excluded?(fact_name)
+    @exclude_list.any? {|excluded_fact| fact_name =~ /#{excluded_fact}/}
```
Comment (joshcooper) [inline lib/puppet/util/fact_dif.rb:55]: Nit this is linear scan for each fact. Might be better to create `@exclude = Set.new(%w[...])` and then check if `@exclude.include?(fact_name)`?

**Comment by joshcooper** [comment]
jenkins please test this

File: lib/puppet/face/facts.rb (hunk 3 above)
Comment (Dorin-Pleava) [reply lib/puppet/face/facts.rb]: Removed from list, as it is not volatile

**Comment by joshcooper** [comment]
Thanks for adding this! I noticed a difference when running in CI. It seems facterng reports the following fact that facter 3.x doesn't. Is that expected (since this is 6.x, puppet defaults to `puppet facts find`, so I have to query in the `.values` hash)?

```
[root@heavy-alchemy ~]# puppet facts . --render-as json --facterng | jq .values.hypervisors
{
  "vmware": {
    "version": "ESXi 6.7"
  }
}
[root@heavy-alchemy ~]# puppet facts . --render-as json | jq .values.hypervisors
{
  "vmware": {
    "version": ""
  }
}
```

**Pull request merged**